infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java

```diff
@@ -0,0 +1,216 @@
+package org.corfudb.infrastructure;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nonnull;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.corfudb.infrastructure.BatchWriterOperation.Type;
+import org.corfudb.infrastructure.log.StreamLog;
+import org.corfudb.protocols.wireprotocol.CorfuPayloadMsg;
+import org.corfudb.protocols.wireprotocol.LogData;
+import org.corfudb.protocols.wireprotocol.RangeWriteMsg;
+import org.corfudb.protocols.wireprotocol.TailsResponse;
+import org.corfudb.protocols.wireprotocol.TrimRequest;
+import org.corfudb.protocols.wireprotocol.WriteRequest;
+import org.corfudb.runtime.exceptions.WrongEpochException;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError;
+
+/**
+ * This class manages access for operations that need ordering while executing against
+ * the backing storage.
+ */
+@Slf4j
+public class BatchProcessor implements AutoCloseable {
```

Comment:
I have a question on the design of BatchProcessor, it is processing requests one by one, but commits (i.e. completes future) and sends RPC responses after a batch, which means there will be a cpu/network IO spike after each batch. Is this intended? Since now the log unit is fully asynchronous, maybe we don't need to batch it?

Reply:
The initial design of the BatchWriter (now called BatchProcessor) is to execute a batch of writes, then fsync the batch. Notice that we can only acknowledge the writes after the fsync completes. That's why the futures are completed only after a successful fsync. I think naturally it is a "spiky" workload: multiple producers and a single consumer. 

Right now, I don't believe this is an issue, but in general it is. We still need to address topic of throttling, but that will come when performance improves.